### ISSUE TITLE
tmpfiles does not prefix argument with root when copying files

### ISSUE THREAD

**Issue by tlahdekorpi** [systemd/systemd#12467]
**systemd version the issue has been seen with**
> 242

**Expected behaviour you didn't see**
 > tmpfiles copying /rootfs/src to /rootfs/dst

**Unexpected behaviour you saw**
 > tmpfiles tries to copy /src

**Steps to reproduce the problem**
`echo 'C /dst - - - - /src' | systemd-tmpfiles --create --root=/rootfs -`

src should be prefixed with the alternate root

**Comment by boucman** [comment]
I am not sure what you mean... --rootfs only changes where systmd-tmpfiles will create the file. it doesn't (and should not) change the content that is written

--rootfs is here for chroot type manipulation (or cross-image creation) so /dst and /src are the correct paths and shouldn't be tweaked (more precisely, they will be the correct path when within the chroot)

Again i'm not sure what you mean, but this sounds like "not a bug" to me...

**Comment by poettering** [comment]
I agree with @boucman: symlink contents is really an opaque string, it doesn't even have to be path at all in fact. Hence we should just pass it through unmodified. If you want the symlink to be affected by the --root= setting I'd recommend generating relative rather than absolute symlinks.

That said, this is definitely something to document clearly and suggest in the documentation, hence marking as a doc issue.

**Comment by tlahdekorpi** [comment]
All paths except the argument for `C` is prefixed with `--root`, this is not expected or desired behavior. Also contradicts what the manpage says.

**Comment by boucman** [comment]
can you give us an examples of "All Paths" ? you mean that if instead of a C line it were a l line, the symlink would be prefixed ?

That sounds like a bug...

**Comment by tlahdekorpi** [comment]
I'm trying to copy files inside the alternate root, not symlink anything anywhere. tmpfiles tries to copy files from outside `--root`.

Both path and argument for `C` (Recursively copy a file or directory) should be treated as files inside the alternate root.

**Comment by boucman** [comment]
that's not how it's supposed to work. The use-case for --root is to build a rootfs within a rootfs, not to automatically add prefix to the tmpfiles line. 

If we change the semantic, we will break the real use-case for --root. I believe --root is not the right tool for what you are trying to do
would
`echo 'C /rootfs/dst - - - - /rootfs/src' | systemd-tmpfiles --create  -`
work for you ?

**Comment by tlahdekorpi** [comment]
It is intended for tmpfiles to use `/usr/share/factory` from outside the `--root`?
e.g. When tmpfiles is being run from initrd to setup booting a system (`systemd-tmpfiles --create --root=/sysroot myconf.conf`) it is supposed to use the factory directory _inside_ the initrd and not the mounted /sysroot?

If so, then this is not a bug. But the documentation should mention this.

**Comment by elboulangero** [comment]
The documentation states:

> --root=root Takes a directory path as an argument. **All paths will be prefixed** with the given alternate root path, including config search paths.

However it seems that in the case of `C` (ie "factory mode"), the path given in argument is not prefixed.

A more detailed example in case it was not clear. Running from dracut. Let's assume I want to copy `/sysroot/usr/libexec` to `/sysroot/var/libexec`. But for some reason I don't want to hardcode `/sysroot` in the tmpfile config file (maybe I use an existing file that I don't want to modify for dracut), so I try to use `--root=`.

```
# echo 'C /var/libexec - - - - /usr/libexec' | systemd-tmpfiles --root=/sysroot --create -
Failed to copy files to /var/libexec: No such file or directory
```
Here, the `no such file or directory` refers to `/usr/libexec`, not to `/var/libexec`.

Now we can pretend that `/usr/libexec` exists and retry:

```
# ln -s /sysroot/usr/libexec /usr/libexec
# echo 'C /var/libexec - - - - /usr/libexec' | systemd-tmpfiles --root=/sysroot --create -
# echo $?
0
# ls /sysroot/var/libexec
   ... mostly flatpak files ...
```

My use-case is a first boot case, I want to initialize a bunch of files super early during the initrd, and I'm trying to do it with systemd-tmpfiles rather than a bash script, basically. I'm even trying NOT TO write a custom tmpfile fragment for this, but instead re-use some that already exist, for example something like:

```
/bin/systemd-tmpfiles --create --root=/sysroot /sysroot/usr/lib/tmpfiles.d/this-file.conf
```

I had the impression that the `--root` argument was meant to do just that, but maybe I'm wrong.

Anyway I just hoped to clarify as I was fighting with the same issue.

**Comment by poettering** [comment]
BTW, you can already do what you are asking for, just use relative symlink/copy paths. They are taken relatively to the destination paths...

```
C /my/path/somewhere - - - - ../../somewhere/else/in/the/same/tree
```

**Comment by poettering** [comment]
Fix waiting in #12686

**Comment by keszybz** [comment]
#12750 has an alternative fix.
